source/python/idetect/fact_api.py

```diff
@@ -206,60 +222,29 @@ def get_wordcloud(session, engine, sample=1000, **filters):
-            FactApi.source_common_name,
-            FactApi.specific_reported_figure,
-            FactApi.term,
-            FactApi.unit,
-            FactApi.vague_reported_figure,
-            FactApi.category,
-            FactApi.gkg_id,
-            FactApi.content_id,
-        ), **filters)
-            .order_by(FactApi.gdelt_day, FactApi.gkg_id)
-            .limit(limit)
-            .offset(offset)
-            .subquery()
-    )
-
-    # find all of the locations for the matching facts. this can't be done in the
-    # query above because there may be some location filters that interfere
-    # note: location names and ids are not always in the same order
-    facts_locations = (
-        session.query(facts,
-                      func.sort(func.array_agg(Location.id)).label('location_ids'),
-                      func.array_agg(Location.location_name).label('location_names'))
-            .join(fact_location, facts.c.fact == fact_location.c.fact)
-            .join(Location, fact_location.c.location == Location.id)
-            .group_by(facts)
-            .subquery()
-    )
-
-    query = (
+    # TODO add one step to verify how many rows are selected,
+    # if the number is larger than a threshold (let's say 50000)
+    # we process only 50000 rows of idetect_fact_api using TABLESAMPLE
+    facts = (add_filters(
         session.query(
-            facts_locations.c.document_identifier.label('document_identifier'),
-            facts_locations.c.fact.label('fact_id'),
-            facts_locations.c.gdelt_day.label('gdelt_day'),
-            facts_locations.c.iso3.label('iso3'),
-            facts_locations.c.source_common_name.label('source_common_name'),
-            facts_locations.c.specific_reported_figure.label('specific_reported_figure'),
-            facts_locations.c.term.label('term'),
-            facts_locations.c.unit.label('unit'),
-            facts_locations.c.vague_reported_figure.label('vague_reported_figure'),
-            facts_locations.c.category.label('category'),
-            facts_locations.c.gkg_id.label('gkg_id'),
-            facts_locations.c.location_ids.label('location_ids'),
-            facts_locations.c.location_names.label('location_names'),
+            FactApi.document_identifier.label('document_identifier'),
+            FactApi.fact.label('fact_id'),
+            FactApi.gdelt_day.label('gdelt_day'),
+            FactApi.iso3.label('iso3'),
+            FactApi.source_common_name.label('source_common_name'),
+            FactApi.specific_reported_figure.label('specific_reported_figure'),
+            FactApi.term.label('term'),
+            FactApi.unit.label('unit'),
+            FactApi.vague_reported_figure.label('vague_reported_figure'),
+            FactApi.category.label('category'),
+            FactApi.gkg_id.label('gkg_id'),
+            FactApiLocations.location_ids.label('location_ids'),
+            FactApiLocations.location_names.label('location_names'),
```

Comment:
I tried actually in a single query in both ways and the cost is the same. It seems to me that the query planner is taking care of the optimisation.

Since urllist is now significantly faster I am also considering having urllist_grouped only returning the groups and call urllist in the application when the user wants to see the facts that are included in the group. This would simplify even further the query

Reply:
Great! I love it when they optimizer does the hard part.

I think it's a great idea to only populate the groups when the user wants to see the facts. The *vast* majority of the groups will never be viewed by a human, so why make the system calculate them?